vm/vm/main/modules/modthread.hh

```diff
@@ -132,6 +132,50 @@ public:
         result = build(vm, "blocked");
     }
   };
+
+  class Suspend: public Builtin<Suspend> {
+  public:
+    Suspend(): Builtin("suspend") {}
+
+    static void call(VM vm, In thread) {
+      Runnable* runnable = getArgument<Runnable*>(vm, thread);
+
+      runnable->suspend();
```

Comment:
This won't be enough if a thread wants to suspend itself. The easiest way to deal with that might be to do the dispatch in Oz, with:
```oz
suspend: proc {$ T}
  if {Boot_Thread.this} == T then
    {Boot_Thread.preempt}
  else
    {Boot_Thread.suspend T}
  end
end
```

Reply:
Ah, indeed. I do not know if suspending itself should preempt or not. I guess it makes sense, but we still have to suspend it. Suspend is like a preempt until resume is called.

Reply:
Hum, you're right, it should definitely *suspend*, not just preempt. But if we only suspend, I think the emulator loop won't notice that it needs to unschedule itself now. It's a bit tricky. Maybe the following works:
```oz
suspend: proc {$ T}
  {Boot_Thread.suspend T}
  if {Boot_Thread.this} == T then
    {Boot_Thread.preempt}
  end
end
```
but maybe that causes all sorts of complications (as some bytecode instructions will keep running even though the thread is marked as suspended.

It might not be possible to do this properly in Oz after all.

Reply:
A thread is never "marked" as suspended. It is just removed from the active queues. suspend + preempt should be perfectly fine.